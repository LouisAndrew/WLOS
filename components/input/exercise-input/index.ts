import ExerciseInput from './exercise-input'

export { ExerciseInput }
