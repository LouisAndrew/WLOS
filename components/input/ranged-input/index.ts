import RangedInput from './ranged-input'

export { RangedInput }
