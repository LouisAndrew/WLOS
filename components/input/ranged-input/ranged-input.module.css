.wrapper {
  @apply flex  flex-col items-center;
}

.input-group {
}

.input {
  @apply text-xl text-center font-display;
}

.input.error {
  @apply text-red-400;
}

.input-separator {
}

.placeholder {
  @apply font-body font-light text-sm;
}
