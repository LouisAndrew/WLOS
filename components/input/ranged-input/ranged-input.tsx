/* eslint-disable object-curly-newline */
import React, { FC, useEffect, useRef, useState } from 'react'
import { parseInt } from 'lodash'
import classname from 'classnames/bind'

import styles from './ranged-input.module.css'

const cx = classname.bind(styles)

export type Props = {
  /**
   * Max number of digits.
   */
  maxDigit: number
  /**
   * Placeholder for the ranged input.
   */
  placeholder?: string
  /**
   * Identifier if the input is editable.
   */
  isEditable?: boolean
  /**
   * Change function whenever value is changed.
   */
  onChange?: (start: number, end?: number) => void
  /**
   * Custom styling for the component.
   */
  className?: string
  /**
   * Default value for start.
   */
  defaultStart?: number
  /**
   * Default value for end.
   */
  defaultEnd?: number
  /**
   * Custom test-id for testing purposes.
   */
  testId?: string
  [key: string]: any
}

const RangedInput: FC<Props> = ({
  maxDigit = 1,
  placeholder = 'PLC',
  onChange = () => {},
  defaultEnd,
  defaultStart,
  className,
  isEditable = false,
  testId = '',
  ...props
}) => {
  const [start, setStart] = useState(defaultStart || -1)
  const [end, setEnd] = useState(defaultEnd || -1)
  const [isSeparatorFilled, setIsSeparatorFilled] = useState(!!defaultStart && !!defaultEnd)

  const [shouldEndShowError, setShouldEndShowError] = useState(false)

  const [shouldSeparatorRender, setShouldSeparatorRender] = useState(!!defaultEnd)
  const [shouldEndRender, setShouldEndRender] = useState(!!defaultEnd)

  const startInput = useRef<HTMLInputElement | null>(null)
  const separatorInput = useRef<HTMLInputElement | null>(null)
  const endInput = useRef<HTMLInputElement | null>(null)

  const lengthValidator = (str: string) => str.length <= maxDigit
  const isNumberValidator = (str: string) => {
    const num = parseInt(str, 10)
    return (!Number.isNaN(num) && num > 0) || str === ''
  }

  const handleChangeStart = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (isEditable) {
      const str = e.target.value
      if (str) {
        if (lengthValidator(str) && isNumberValidator(str)) {
          setStart(parseInt(str))
          if (str.length === maxDigit) {
            separatorInput.current?.focus()
          }
        }
      } else {
        setStart(-1)
      }
    }
  }

  const handleKeyDownStart = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (isEditable) {
      if (e.key === 'Enter') {
        if (start !== -1) {
          separatorInput.current?.focus()
        }
      }
    }
  }

  const handleChangeSeparator = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (isEditable) {
      if (start === -1) {
        startInput.current.focus()
        return
      }
      const value = e.target.value === '-'
      setIsSeparatorFilled(value)
      if (value) {
        endInput.current?.focus()
      }
    }
  }

  const handleKeyDownSeparator = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (isEditable) {
      if (e.key === 'Backspace') {
        if (isSeparatorFilled) {
          if (end === -1) {
            setIsSeparatorFilled(false)
          }
        } else {
          startInput.current?.focus()
        }
      }

      if (e.key === 'Enter') {
        if (start !== -1) {
          setIsSeparatorFilled(true)
          endInput.current?.focus()
          return
        }

        startInput.current?.focus()
      }
    }
  }

  const handleChangeEnd = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (isEditable) {
      if (start === -1) {
        startInput.current?.focus()
        return
      }

      if (!isSeparatorFilled) {
        separatorInput.current?.focus()
        return
      }

      const str = e.target.value
      if (str) {
        if (lengthValidator(str) && isNumberValidator(str)) {
          setEnd(parseInt(str))
        }
      } else {
        setEnd(-1)
      }
    }
  }

  const handleKeyDownEnd = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (isEditable) {
      if (e.key === 'Backspace') {
        if (end === -1) {
          separatorInput?.current?.focus()
        } else {
          const endToString = end.toString()
          if (endToString.length === 1) {
            setEnd(-1)
            return
          }

          setEnd(parseInt(endToString.substring(0, endToString.length - 1)))
        }
      }

      if (e.key === 'Enter') {
        if (end !== -1 && start !== -1 && isSeparatorFilled) {
          endInput.current?.blur()
        }
      }
    }
  }

  const renderAll = () => {
    setShouldEndRender(true)
    setShouldSeparatorRender(true)
  }

  const collapseUnecessary = () => {
    if (end === start) {
      setShouldEndRender(false)
      setShouldSeparatorRender(false)
    }
  }

  useEffect(() => {
    setShouldEndShowError(end < start)
  }, [end, start])

  useEffect(() => {
    const shouldCallWithEnd = end !== -1 && end > start
    if (start !== -1) {
      onChange(start, shouldCallWithEnd ? end : undefined)
    }

    if (start !== -1 && end !== -1) {
      setIsSeparatorFilled(true)
    }
  }, [start, end])

  const widthMultiplier = 6
  const width = 4
  const inputPlaceholder = Array.from(new Array(maxDigit).keys())
    .map(() => '#')
    .join('')

  const inputSeparatorClass = cx({
    input: true,
    'input-separator': true,
  })

  const endInputClass = cx({
    input: true,
    error: shouldEndShowError,
  })

  return (
    <div
      className={`${styles.wrapper} ${className}`}
      data-testid="ranged-input-wrapper"
      onFocus={() => {
        if (isEditable) {
          renderAll()
        }
      }}
      onBlur={() => collapseUnecessary()}
      {...props}
    >
      <div className={styles['input-group']}>
        <input
          ref={startInput}
          value={start === -1 ? '' : start}
          onChange={handleChangeStart}
          onKeyDown={handleKeyDownStart}
          className={styles.input}
          style={{ width: widthMultiplier * maxDigit * width }}
          placeholder={inputPlaceholder}
          required
          aria-label="Range start"
          data-testid={`${testId}-range-start`}
        />
        <input
          ref={separatorInput}
          value={isSeparatorFilled ? '-' : ''}
          placeholder="-"
          onChange={handleChangeSeparator}
          onKeyDown={handleKeyDownSeparator}
          className={inputSeparatorClass}
          style={{
            width: shouldSeparatorRender ? widthMultiplier * width : 0,
            opacity: shouldSeparatorRender ? 1 : 0,
          }}
          aria-label="Range separator"
          data-testid={`${testId}-range-separator`}
        />
        <input
          value={end === -1 ? '' : end}
          ref={endInput}
          onChange={handleChangeEnd}
          onKeyDown={handleKeyDownEnd}
          className={endInputClass}
          style={{
            width: shouldEndRender ? widthMultiplier * maxDigit * width : 0,
            opacity: shouldEndRender ? 1 : 0,
          }}
          placeholder={inputPlaceholder}
          aria-label="Range end"
          data-testid={`${testId}-range-end`}
        />
      </div>
      <div className={styles.placeholder}>{placeholder.toUpperCase()}</div>
    </div>
  )
}

export default RangedInput
