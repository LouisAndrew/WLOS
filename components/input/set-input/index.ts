import SetInput from './set-input'

export { SetInput }