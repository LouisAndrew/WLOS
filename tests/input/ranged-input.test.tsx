import { render, cleanup, fireEvent } from '@testing-library/react'
import { RangedInput } from '@c/input/ranged-input'
afterEach(cleanup)

const onChange = jest.fn()
const Component = <RangedInput maxDigit={1} onChange={onChange} isEditable={true} />

const labels = {
  start: 'Range start',
  separator: 'Range separator',
  end: 'Range end',
}

// todo: write tests for keymaps
describe('Ranged input', () => {
  it('matches snapshot', () => {
    expect(render(Component)).toMatchSnapshot()
  })

  it('hides separator and range-end inputs on default', () => {
    const { getByLabelText: q } = render(Component)
    expect(q(labels.end)).not.toBeVisible()
    expect(q(labels.separator)).not.toBeVisible()
  })

  it('displays separator and range-end inputs when input wrapper is focused', () => {
    const { getByLabelText: q, getByTestId: qt } = render(Component)
    fireEvent.focus(qt('ranged-input-wrapper'))
    expect(q(labels.end)).toBeVisible()
    expect(q(labels.separator)).toBeVisible()
  })

  it('focuses the next input field if the current field is filled with numbers of character specified by `maxDigit` prop', () => {
    const { getByLabelText: q } = render(Component)
    fireEvent.change(q(labels.start), { target: { value: '1' } })
    expect(q(labels.separator)).toHaveFocus()

    fireEvent.change(q(labels.separator), { target: { value: '-' } })
    expect(q(labels.end)).toHaveFocus()

    fireEvent.change(q(labels.end), { target: { value: '2' } })
    expect(q(labels.end)).toHaveFocus()
  })

  it('does not change value of input field if the predecessor field has not filled yet', () => {
    const { getByLabelText: q } = render(Component)
    const start = q(labels.start)
    const end = q(labels.end)
    const separator = q(labels.separator)

    fireEvent.change(separator, { target: { value: '-' } })
    expect(separator.getAttribute('value')).not.toBe('-')
    expect(start).toHaveFocus()

    fireEvent.change(end, { target: { value: '2' } })
    expect(end.getAttribute('value')).not.toBe('2')
    expect(start).toHaveFocus()
  })

  it('sets the range-end input state to error if the end value is smaller than start value', () => {
    const { getByLabelText: q } = render(Component)
    const start = q(labels.start)
    const end = q(labels.end)

    fireEvent.change(start, { target: { value: '3' } })
    fireEvent.change(q(labels.separator), { target: { value: '-' } })
    fireEvent.change(end, { target: { value: '2' } })
    expect(end.classList).toContain('error')
  })
})
