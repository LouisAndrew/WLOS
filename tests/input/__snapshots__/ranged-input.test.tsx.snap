// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ranged input matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="wrapper undefined"
        data-testid="ranged-input-wrapper"
      >
        <div
          class="input-group"
        >
          <input
            aria-label="Range start"
            class="input"
            data-testid="-range-start"
            placeholder="#"
            required=""
            style="width: 24px;"
            value=""
          />
          <input
            aria-label="Range separator"
            class="input input-separator"
            data-testid="-range-separator"
            placeholder="-"
            style="width: 0px; opacity: 0;"
            value=""
          />
          <input
            aria-label="Range end"
            class="input"
            data-testid="-range-end"
            placeholder="#"
            style="width: 0px; opacity: 0;"
            value=""
          />
        </div>
        <div
          class="placeholder"
        >
          PLC
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="wrapper undefined"
      data-testid="ranged-input-wrapper"
    >
      <div
        class="input-group"
      >
        <input
          aria-label="Range start"
          class="input"
          data-testid="-range-start"
          placeholder="#"
          required=""
          style="width: 24px;"
          value=""
        />
        <input
          aria-label="Range separator"
          class="input input-separator"
          data-testid="-range-separator"
          placeholder="-"
          style="width: 0px; opacity: 0;"
          value=""
        />
        <input
          aria-label="Range end"
          class="input"
          data-testid="-range-end"
          placeholder="#"
          style="width: 0px; opacity: 0;"
          value=""
        />
      </div>
      <div
        class="placeholder"
      >
        PLC
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
